src/nvim/lua/vim.lua

```diff
@@ -118,11 +118,91 @@ local function _update_package_paths()
   last_nvim_paths = cur_nvim_paths
 end
 
+local function gsplit(s, sep, plain)
+  assert(type(s) == "string")
+  assert(type(sep) == "string")
+  assert(type(plain) == "boolean" or type(plain) == "nil")
+
+  local start = 1
+  local done = false
+
+  local function pass(i, j, ...)
+    if i then
+      assert(j+1 > start, "Infinite loop detected")
+      local seg = s:sub(start, i - 1)
+      start = j + 1
+      return seg, ...
+    else
+      done = true
+      return s:sub(start)
+    end
+  end
+
+  return function()
+    if done then
+      return
+    end
+    if sep == '' then
+      if start == #s then
+        done = true
+      end
+      return pass(start+1, start)
+    end
+    return pass(s:find(sep, start, plain))
+  end
+end
+
+local function split(s,sep,plain)
+  local t={} for c in gsplit(s, sep, plain) do table.insert(t,c) end
+  return t
+end
+
+local function trim(s)
+  assert(type(s) == "string", "Only strings can be trimmed")
+  local result = s:gsub("^%s+", ""):gsub("%s+$", "")
+  return result
+end
+
+local function deepcopy(orig)
+    local copy
+    if type(orig) == 'table' then
+        copy = {}
+        for k, v in pairs(orig) do
+            copy[deepcopy(k)] = deepcopy(v)
+        end
+        setmetatable(copy, deepcopy(getmetatable(orig)))
+    else -- number, string, boolean, etc
+        copy = orig
+    end
+    return copy
+end
+
+local deepcopy_funcs = {
+  table = function(orig)
+    local copy = {}
+    for k, v in pairs(orig) do
+      copy[deepcopy(k)] = deepcopy(v)
+    end
+    return copy
+  end,
+  number = id,
+  string = id,
+  ['nil'] = id,
+  boolean = id,
+}
+
+local deepcopy = function(orig)
```

Comment:
we always use the `local function foo()` form. (There might be some old code that needs to be updated)

Reply:
The local needs to be removed, but otherwise it's needed this way, because we need to declare `deepcopy` before `deepcopy_funcs`.